src/source/Crypto/Dtls_openssl.c

```diff
@@ -193,12 +245,16 @@ STATUS createSslCtx(PDtlsSessionCertificateInfo pCertificates, UINT32 certCount,
 
     CHK(pSslCtx != NULL, STATUS_SSL_CTX_CREATION_FAILED);
 
-    // Version greater than or equal to 1.0.2
-#if (OPENSSL_VERSION_NUMBER >= 0x10002000L)
-    SSL_CTX_set_ecdh_auto(pSslCtx, TRUE);
-#else
+// https://www.openssl.org/docs/man1.0.2/man3/
+// https://www.openssl.org/docs/man1.1.1/man3/
+// Version < 3.0.0
+#if (OPENSSL_VERSION_NUMBER < 0x30000000L)
     CHK((ecdh = EC_KEY_new_by_curve_name(NID_X9_62_prime256v1)) != NULL, STATUS_SSL_CTX_CREATION_FAILED);
     CHK(SSL_CTX_set_tmp_ecdh(pSslCtx, ecdh) == 1, STATUS_SSL_CTX_CREATION_FAILED);
+#else
+    // https://www.openssl.org/docs/man3.0/man3/EC_KEY_new_by_curve_name.html -- indicates that EC_KEY_new_by_curve_name and SSL_CTX_set_tmp_ecdh are
+    // deprecated https://www.openssl.org/docs/man1.1.1/man3/SSL_CTX_set1_groups.html
+    CHK(SSL_CTX_set1_groups_list(pSslCtx, "prime256v1") == 1, STATUS_SSL_CTX_CREATION_FAILED);
```

Comment:
Ok so I see that you're using the new API `SSL_CTX_set1_groups_list` for open ssl >= 3.0.0.  But where the diff now is compared to our existing code is for all versions higher than 1.0.2 (so 1.0.2 <= x < 3.0.0), really we did not enforce the upper bound before but for all intents in purposes for theses versions we were previously using `SSL_CTX_set_ecdh_auto`.  That I want to understand why are we changing the behavior for existing versions of openssl that's already been tried and tested, was there a bug with this or what is the reason for this change?  IMO unless there is a good reason this change should only impact when the version is >= 3.0.0 and otherwise it should not touch the existing behavior at all, unless there is a good reason to do that.  I'm not following what that reason is.

Reply:
`SSL_CTX_set_ecdh_auto` is primarily used for automatic curve selection during SSL/TLS handshakes, making it convenient for situations where you want to delegate curve selection to OpenSSL. On the other hand, `EC_KEY_new_by_curve_name` is used when you need to explicitly specify the ECC curve for ECC key generation and want more control over the curve selection process. I am not sure what `SSL_CTX_set_ecdh_auto` selects by default, but given we are controlling the curve selection pre 1.0.2, it only made sense to do the same post 1.0.2 since the API is available. Another reason for the change was I wanted to avoid having multiple ifdefs that make the code buggy when it is unnecessary. I do understand the point of the existing code already be tested, but the functions I have included for versions between 1.0.2 and 3.0.0 are also tested as being part of version <1.0.2. Without this change, the ifdefs look like:

```
#if (OPENSSL_VERSION_NUMBER >= 0x10002000L && OPENSSL_VERSION_NUMBER < 0x3000000L)
    SSL_CTX_set_ecdh_auto(pSslCtx, TRUE);
#elif (OPENSSL_VERSION_NUMBER < 0x10002000L)
    CHK((ecdh = EC_KEY_new_by_curve_name(NID_X9_62_prime256v1)) != NULL, STATUS_SSL_CTX_CREATION_FAILED);
    CHK(SSL_CTX_set_tmp_ecdh(pSslCtx, ecdh) == 1, STATUS_SSL_CTX_CREATION_FAILED);
#elif (OPENSSL_VERSION_NUMBER >= 0x3000000L)
    CHK(SSL_CTX_set1_groups_list(pSslCtx, "prime256v1") == 1, STATUS_SSL_CTX_CREATION_FAILED);
#endif
```
Let me know if you still feel strongly about this change and I can switch it to this block. But, given `EC_KEY_new_by_curve_name` and `SSL_CTX_set_tmp_ecdh` are already tested for versions less than 1.0.2, I did not see the harm in using the same for 1.0.2 to 3.0.0.

Reply:
I did some digging and I found `SSL_CTX_set_ecdh_auto` was entirely removed in openssl 1.1.0 [https://github.com/openssl/openssl/issues/1437#issuecomment-238828306].  Then it was added back as a dummy no-op macro for backwards compat like a week later in 1.1.0a (https://github.com/openssl/openssl/commit/2ecb9f2d18614fb7b7b42830a358b7163ed43221).  So I suppose we've actually been getting lucky if someone was using this against 1.1.0 it wouldn't compile.

I cannot seem to find the manpages for openssl 1.0.0 to verify if `EC_KEY_new_by_curve_name` existed then, and also what about the curve itself, has `NID_X9_62_prime256v1` always existed.  Let me just see if I can find some official references if not I'll look through the source.

Reply:
Yeah. There doesnt seem to be official man pages pre-1.0.2. But I just checked out the branch to check if the API existed.